### PR TITLE
[pytorch] change detach() & detach_() to no-op for USE_STATIC_DISPATCH mode

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28402 [pytorch] remove AutoNonVariableTypeMode from jit-op-registry
* **#28400 [pytorch] change detach() & detach_() to no-op for USE_STATIC_DISPATCH mode**
* #28399 [pytorch] remove AutoNonVariableTypeMode guard around forward() call
* #28398 [pytorch] Make static dispatch turn off variable before entering the kernel

Summary:
This is yet-another fix to issue #26764.

Some mobile models call tensor.detach() which won't work with
static-dispatch mode. We disable autograd for static-dispatch / mobile
build anyway so it seems fine to make these op-ops.

Test Plan:
- With stacked PRs, confirmed it can run failed models now.

Differential Revision: [D18055852](https://our.internmc.facebook.com/intern/diff/D18055852)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by gchanan** [comment]
so, what's the plan for addressing this when STATIC_DISPATCH goes away?
